src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -3585,6 +3585,21 @@ static QVariant fcnSinuosity( const QVariantList &values, const QgsExpressionCon
   return QVariant( curve->sinuosity() );
 }
 
+static QVariant fcnStraightDistance2d( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  const QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+  const QgsCurve *curve = qgsgeometry_cast< const QgsCurve * >( geom.constGet() );
```

Comment:
```suggestion
  const QgsCurve *curve = geom.constGet() ? qgsgeometry_cast< const QgsCurve * >( geom.constGet()->simplifiedTypeRef() ) : nullptr;
```

There's a special situation which needs handling here, which is that the input geometry is a MultiLineString geometry containing only a single part. This happens a lot, e.g. if the input layer is a shapefile then all geometries will be reported as multilinestrings, even if they only have a single line part. So it's something we need to also allow.

Calling QgsAbstractGeometry::simplifiedTypeRef() is the easiest/fastest way to handle this. simplifiedTypeRef will return either the original geometry if it's already a single-line type, OR the first linestring part if it's a multilinestring containing a single part, OR the original geometry ref if it's neither of these (i.e. a multilinestring with multiple parts). So ultimately it's a really quick way of handling the "single line part in a multilinestring" case without involving any copying of geometries!